subsys/net/lib/ocpp/ocpp.c

```diff
@@ -0,0 +1,739 @@
+
+			rcv.ctx = ctx;
+			rcv.msg = ui->recv_buf;
+			rcv.sndlock = &ui->ws_sndlock;
+			rcv.msg_len = CONFIG_OCPP_RECV_BUFFER_SIZE;
+
+			memset(ui->recv_buf, 0, rcv.msg_len);
+			ret = ocpp_receive_from_server(&rcv, 200);
+			if (ret < 0) {
+				continue;
+			}
+
+			ocpp_process_server_msg(ctx);
+		}
+	}
+}
+
+int ocpp_upstream_init(ocpp_info_t *ctx, ocpp_cs_info_t *csi)
+{
+	int ret;
+	ocpp_upstream_info_t *ui;
+
+	LOG_INF("upstream init");
+
+	ui = k_calloc(1, sizeof(ocpp_upstream_info_t));
+	if (!ui) {
+		return -ENOMEM;
+	}
+
+	ui->wsrecv_buf = k_calloc(1, CONFIG_OCPP_RECV_BUFFER_SIZE * 2);
+	if (!ui->wsrecv_buf) {
+		ret = -ENOMEM;
+		goto out;
+	}
+
+	ui->recv_buf = k_calloc(1, CONFIG_OCPP_RECV_BUFFER_SIZE);
+	if (!ui->recv_buf) {
+		ret = -ENOMEM;
+		goto out;
+	}
+
+	ui->csi.ws_url = strdup(csi->ws_url);
+	ui->csi.cs_ip = strdup(csi->cs_ip);
+	ui->csi.port = csi->port;
+	ui->csi.sa_family = csi->sa_family;
+	ui->tcpsock = zsock_socket(csi->sa_family, SOCK_STREAM,
+				   IPPROTO_TCP);
+	if (ui->tcpsock < 0) {
+		ret = -errno;
+		goto out;
+	}
+
+	k_mutex_init(&ui->ws_sndlock);
+	k_poll_signal_init(&ui->ws_rspsig);
+	ui->wssock = -1;
+	ctx->ui = ui;
+
+	websocket_init();
+
+	k_thread_create(&ui->tinfo, ocpp_wsreader_stack,
+			CONFIG_OCPP_WSREADER_THREAD_STACKSIZE,
+			ocpp_wsreader, ctx, NULL, NULL,
+			OCPP_UPSTREAM_PRIORITY,	0, K_MSEC(100));
+
+	return 0;
+
+out:
+	if (ui) {
+		k_free(ui->wsrecv_buf);
+		k_free(ui->recv_buf);
+	}
+	k_free(ui);
+
+	return ret;
+}
+
+static void timer_heartbeat_cb(struct k_timer *t)
+{
+	ocpp_info_t *ctx;
+	internal_msg_t msg = {PDU_HEARTBEAT};
+
+	ctx = k_timer_user_data_get(t);
+	if (ctx->state == CP_STATE_BOOTNOTIF) {
+		msg.msgtype = PDU_BOOTNOTIFICATION;
+	}
+
+	k_msgq_put(ctx->msgq, &msg, K_NO_WAIT);
+}
+
+static void timer_meter_cb(struct k_timer *t)
+{
+	ocpp_info_t *ctx;
+	internal_msg_t msg = {PDU_METER_VALUES};
+
+	ctx = k_timer_user_data_get(t);
+
+	k_msgq_put(ctx->msgq, &msg, K_NO_WAIT);
+}
+
+static inline void bootnotification_fill_resource(ocpp_cp_info_t *cp,
+						  ocpp_cp_info_t *cpi,
+						  ocpp_info_t *ctx)
+{
+	struct k_timer *hbt = &ctx->hb_timer;
+
+	cp->model = strdup(cpi->model);
+	cp->vendor = strdup(cpi->vendor);
+	if (cp->sl_no) {
+		cp->sl_no = strdup(cpi->sl_no);
+	}
+
+	if (cp->box_sl_no) {
+		cp->box_sl_no = strdup(cpi->box_sl_no);
+	}
+
+	if (cp->fw_ver) {
+		cp->fw_ver = strdup(cpi->fw_ver);
+	}
+
+	if (cp->iccid) {
+		cp->iccid = strdup(cpi->iccid);
+	}
+
+	if (cp->imsi) {
+		cp->imsi = strdup(cpi->imsi);
+	}
+
+	if (cp->meter_sl_no) {
+		cp->meter_sl_no = strdup(cpi->meter_sl_no);
+	}
+
+	if (cp->meter_type) {
+		cp->meter_type = strdup(cpi->meter_type);
+	}
+
+	ctx->state = CP_STATE_BOOTNOTIF;
+	ctx->hb_sec = 10;
+	k_timer_start(hbt, K_SECONDS(1), K_NO_WAIT);
+}
+
+int ocpp_session_open(ocpp_session_handle_t *hndl)
+{
+	ocpp_session_t *sh;
+
+	if (!hndl)
+		return -EINVAL;
+
+	sh = (ocpp_session_t *)k_calloc(1, sizeof(ocpp_session_t));
```

Comment:
We tend to avoid heap usage in the networking libraries unless it's really unavoidable. Wouldn't it make sense to allow the app to preallocate the session structure, and provide the pointer as an argument to the function?

Reply:
ocpp_session_handle_t exposed to user is opaque with void *, otherwise struct ocpp_session has to be made public has app has nothing with internal members also app should care about allocate/static memory. I would prefer to keep it, suggestion please